src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+
+        """
+        if isinstance(grid, (pp.BoundaryGrid, pp.MortarGrid)):
+            # NOTE using default value of 1, because this is the general default value
+            # and only cells are supported on boundaries and interfaces
+            return self._dof_info.get("cells", 1) * grid.num_cells
+        elif isinstance(grid, pp.Grid):
+            # NOTE cannot use default value of scalar, cell-wise, to not mess with
+            # cases where the user defines only node- or face-wise dofs.
+            return (
+                self._dof_info.get("cells", 0) * grid.num_cells
+                + self._dof_info.get("faces", 0) * grid.num_faces
+                + self._dof_info.get("nodes", 0) * grid.num_nodes
+            )
+        else:
+            raise TypeError(f"Unsupported type of grid {type(grid)}.")
+
+    @property
+    def name(self) -> str:
+        """Name of this expression assigned at instantiation."""
+        return self._name
+
+    @property
+    def _name_diffs(self) -> str:
+        """Private property giving the name under which the **derivative** values of
+        this expression are stored in the grid data dictionaries"""
+        return f"{self.name}_DERIVATIVES"
+
+    # Convenience properties/methods to access and progress values collectively
+
+    @property
+    def boundary_values(self) -> np.ndarray:
+        """Property to access and set the boundary value at the current time step,
+        on **all** boundaries in the md grid.
+
+        The getter fetches the values in the order imposed by the md-grid.
+        The setter slices the values according to :meth:`num_dofs_on_grid` for each
+        boundary grid in the md-grid.
+
+        Note:
+            This is a convenience functionality for :meth:`update_boundary_values`,
+            which operates on all boundary grids.
+
+            Hence, the setter shifts **time step** values backwards and sets the given
+            value as the **curent iterate** (current time).
+
+            Most importantly, it has a different paradigm than the property setters and
+            getters for subdomains and interfaces, which operate only with iterate
+            values.
+
+        Parameters:
+            val: A new value with shape ``(N,)`` to be set. Note that ``N`` is
+                calculated from the information passed in ``dof_info`` during
+                instantiation, and the number of cells in all boundary grids.
+
+        Returns:
+            If boundary values are set, it returns the one for the current time
+            (iterate index = 0). The shape is consistent with ``val``.
+
+        """
+        vals = []
+        for _, data in self.mdg.boundaries(return_data=True):
+            vals.append(pp.get_solution_values(self.name, data, iterate_index=0))
+        if len(vals) > 0:
+            return np.hstack(vals)
+        else:
+            return np.zeros(0, dtype=float)
```

Comment:
Will this give an array of dimension 0 or 1, or a numpy scalar? I very much prefer 1d-arrays, and would really like to avoid numpy scalars, as these bring all kind of corner cases to the parsing (don't ask)

Reply:
It's an instance of `np.ndarray`, with `ndim=1`, `shape=(0,)`, definitely not a scalar.